### PR TITLE
query library records lazily during install

### PR DESCRIPTION
**Pull request by kevinushey**
Closes https://github.com/rstudio/renv/issues/1469.

### CODE REVIEW

File: R/retrieve.R [hunk 1]
```diff
@@ -99,11 +96,9 @@ renv_retrieve_impl <- function(package) {
     is.null(record$Version)
 
   if (uselatest) {
-    record <- renv_available_packages_latest(package)
-    if (is.null(record)) {
+    record <- catch(renv_available_packages_latest(package))
+    if (inherits(record, "error"))
       stopf("package '%s' is not available", package)
```
Comment (hadley) [inline R/retrieve.R:101]:
```R
record <- withCallingHandlers(
  renv_available_packages_latest(package),
  function(err) stopf("package '%s' is not available", package)
)
```
?

File: R/retrieve.R [hunk 2]
```diff
@@ -209,7 +204,10 @@ renv_retrieve_impl <- function(package) {
 
   }
 
-  state$downloaded <- TRUE
+  if (!identical(state$downloaded, TRUE)) {
```
Comment (hadley) [inline R/retrieve.R:207]: This is a nice improvement 😄

File: R/snapshot.R [hunk 3]
```diff
@@ -681,7 +681,12 @@ renv_snapshot_library_diagnose_missing_description <- function(library, pkgs) {
 renv_snapshot_description <- function(path = NULL, package = NULL) {
 
   # read DESCRIPTION file
-  path <- path %||% renv_package_find(package)
+  path <- path %||% {
+    path <- renv_package_find(package)
+    if (!nzchar(path))
+      stopf("package '%s' is not installed", package)
+  }
+
```
Comment (hadley) [inline R/snapshot.R:689]: Do you want to check the path exists even if it's user supplied?

**Review by hadley — APPROVED** [review]

File: R/snapshot.R (hunk 3 above)
Comment (kevinushey) [reply R/snapshot.R:689]: Good catch!

**Pull request merged**